cmd/query/main.go

```diff
@@ -68,13 +82,31 @@ func main() {
 		app.HandlerOptions.Prefix(*builder.QueryPrefix),
 		app.HandlerOptions.Logger(logger))
 	sHandler := app.NewStaticAssetsHandler(*builder.QueryStaticAssets)
+
 	r := mux.NewRouter()
 	rHandler.RegisterRoutes(r)
 	sHandler.RegisterRoutes(r)
-	portStr := ":" + strconv.Itoa(*builder.QueryPort)
+
+	httpPortStr := ":" + strconv.Itoa(*builder.QueryPort)
 	recoveryHandler := recoveryhandler.NewRecoveryHandler(logger, true)
-	logger.Info("Starting jaeger-query HTTP server", zap.Int("port", *builder.QueryPort))
-	if err := http.ListenAndServe(portStr, recoveryHandler(r)); err != nil {
-		logger.Fatal("Could not launch service", zap.Error(err))
+	logger.Info("Starting Jaeger Query HTTP server", zap.Int("port", *builder.QueryPort))
+
+	go func() {
+		// there's still a chance that a request comes in between the `ready` flag and the server's
+		// actual startup, but this would be a rare situation, as we assume the `ListenAndServe` is fast enough
+		// between allocating a port and serving a request.
+		state = http.StatusNoContent
```

Comment:
I ended up implementing as channels, as we are talking about a go channel calling a method on an object set by another go channel. When developing, I got a panic in one situation, as the handler was still null when Update was called. With channels, it works as expected.

Query main HTTP server started before:
```json
{"level":"info","ts":1501236900.1643984,"caller":"query/main.go:56","msg":"Ticking"}
{"level":"info","ts":1501236900.187948,"caller":"query/main.go:99","msg":"Starting Jaeger Query HTTP server","port":16686}
{"level":"info","ts":1501236901.164993,"caller":"healthcheck/handler.go:48","msg":"Health Check server started","http-port":16687}
{"level":"info","ts":1501236901.1652951,"caller":"healthcheck/handler.go:68","msg":"Health Check state change","new-state":204}
```

Health check HTTP server started before:
```json
{"level":"info","ts":1501236874.302359,"caller":"healthcheck/handler.go:49","msg":"Health Check server started","http-port":16687}
{"level":"info","ts":1501236874.3137684,"caller":"query/main.go:94","msg":"Starting Jaeger Query HTTP server","port":16686}
{"level":"info","ts":1501236875.302688,"caller":"healthcheck/handler.go:72","msg":"Health Check state change","new-state":204}
```

PS: I tested it manually with a ticker:

```go
tick := time.Tick(1 * time.Second)
logger.Info("Ticking")
<-tick
```